### PR TITLE
Batch folly::preadv Calls in LocalReadFile::preadv to IOV_MAX

### PR DESCRIPTION
**Pull request by acvictor**
Batch calls to folly::preadv when `iovecs.size() >= IOV_MAX` to conform with limits placed on iovcnt in the underlying system's preadv. Fixes the issue mentioned in https://github.com/facebookincubator/velox/issues/6648.

### CODE REVIEW

File: velox/common/file/File.cpp [hunk 1]
```diff
@@ -142,21 +142,60 @@ uint64_t LocalReadFile::preadv(
   // Dropped bytes sized so that a typical dropped range of 50K is not
   // too many iovecs.
   static thread_local std::vector<char> droppedBytes(16 * 1024);
+  uint64_t totalBytesRead = 0;
   std::vector<struct iovec> iovecs;
   iovecs.reserve(buffers.size());
+
+  auto preadvRun = [&]() -> ssize_t {
```
Comment (xiaoxmeng) [inline velox/common/file/File.cpp:149]:
```
s/preadvRun/readvFunc/
```

File: velox/common/file/File.cpp [hunk 2]
```diff
@@ -142,21 +142,60 @@ uint64_t LocalReadFile::preadv(
...
+    auto bytesRead = folly::preadv(fd_, iovecs.data(), iovecs.size(), offset);
```
Comment (xiaoxmeng) [inline velox/common/file/File.cpp:150]:
```
const auto bytesRead
```

File: velox/common/file/File.cpp [hunk 3]
```diff
@@ -142,21 +142,60 @@ uint64_t LocalReadFile::preadv(
...
+    if (bytesRead < 0) {
+      perror("preadv");
+      LOG(ERROR) << "preadv failed with error: " << strerror(errno);
```
Comment (xiaoxmeng) [inline velox/common/file/File.cpp:153]: Remove perror("preadv");? Is it required?
```
LOG(ERROR) << "preadv failed with error: " << folly::errnoStr(errno);
```

File: velox/common/file/File.cpp [hunk 4]
```diff
@@ -142,21 +142,60 @@ uint64_t LocalReadFile::preadv(
...
+  };
+
   for (auto& range : buffers) {
     if (!range.data()) {
       auto skipSize = range.size();
       while (skipSize) {
         auto bytes = std::min<size_t>(droppedBytes.size(), skipSize);
+
+        if (iovecs.size() >= IOV_MAX) {
+          auto bytesRead = preadvRun();
```
Comment (xiaoxmeng) [inline velox/common/file/File.cpp:169]: Consider add a macro in this file
```
#define RETURN_IF_ERROR(func, result) 
do {
  result = func;
  if (result < 0) {
     return result;
  }
} while (0)
```

File: velox/dwio/common/tests/ReadFileInputStreamTests.cpp [hunk 5]
```diff
@@ -73,3 +75,36 @@ TEST(ReadFileInputStream, VReadIOBufs) {
   std::vector<std::string> expected = {"aaaaab", "bcccc"};
   EXPECT_EQ(result, expected);
 }
+
+TEST(ReadFileInputStream, LocalReadFile) {
```
Comment (xiaoxmeng) [inline velox/dwio/common/tests/ReadFileInputStreamTests.cpp:79]: Consider to add test class for ReadFileInputStreamTest with TEST_F and then handle local file system registration in setup call?

File: velox/dwio/common/tests/ReadFileInputStreamTests.cpp [hunk 6]
```diff
@@ -73,3 +75,36 @@ TEST(ReadFileInputStream, VReadIOBufs) {
...
+  const auto& filename = tempFile->path;
+  remove(filename.c_str());
+  {
+    LocalWriteFile writeFile(filename);
+    for (int i = 0; i < 1027; ++i)
+      writeFile.append("abc");
+    ASSERT_EQ(writeFile.size(), static_cast<uint64_t>(3081));
+  }
+
+  auto file = std::make_shared<LocalReadFile>(filename);
```
Comment (xiaoxmeng) [inline velox/dwio/common/tests/ReadFileInputStreamTests.cpp:91]:
```
s/file/readFile/
```

File: velox/dwio/common/tests/ReadFileInputStreamTests.cpp [hunk 7]
```diff
@@ -73,3 +75,36 @@ TEST(ReadFileInputStream, VReadIOBufs) {
...
+  ASSERT_EQ(static_cast<uint64_t>(3081), file->size());
+
+  auto readFile =
```
Comment (xiaoxmeng) [inline velox/dwio/common/tests/ReadFileInputStreamTests.cpp:94]:
```
s/readFile/readStream/
```

File: velox/dwio/common/tests/ReadFileInputStreamTests.cpp [hunk 8]
```diff
@@ -73,3 +75,36 @@ TEST(ReadFileInputStream, VReadIOBufs) {
...
+      std::make_shared<facebook::velox::dwio::common::ReadFileInputStream>(
+          file);
+  std::vector<Region> regions;
```
Comment (xiaoxmeng) [inline velox/dwio/common/tests/ReadFileInputStreamTests.cpp:97]:
```

std::vector<Region> regions;
for (int i = 0; i < 3081; i += 3) {
```

**Review by xiaoxmeng — COMMENTED** [review]
@acvictor thanks for the change % minors

File: velox/dwio/common/tests/ReadFileInputStreamTests.cpp [hunk 9]
```diff
@@ -73,3 +75,36 @@ TEST(ReadFileInputStream, VReadIOBufs) {
...
+        Region(static_cast<uint64_t>(i), static_cast<uint64_t>(1)));
+  }
+  ASSERT_GT(regions.size(), IOV_MAX);
+
+  std::vector<folly::Range<char*>> buffers;
+  std::vector<folly::IOBuf> iobufs(regions.size());
+
+  readFile->vread(regions, {iobufs.data(), iobufs.size()}, LogType::TEST);
+  remove(filename.c_str());
+}
```
[the remark this answers is not in the document]
Comment (acvictor) [reply velox/dwio/common/tests/ReadFileInputStreamTests.cpp]: Done.

File: velox/common/file/File.cpp (hunk 2 above)
Comment (acvictor) [reply velox/common/file/File.cpp]: Updated.

File: velox/common/file/File.cpp (hunk 4 above)
Comment (acvictor) [reply velox/common/file/File.cpp]: Thanks for the suggestion! Updated.

File: velox/dwio/common/tests/ReadFileInputStreamTests.cpp (hunk 7 above)
Comment (acvictor) [reply velox/dwio/common/tests/ReadFileInputStreamTests.cpp]: Renamed.

File: velox/dwio/common/tests/ReadFileInputStreamTests.cpp (hunk 8 above)
Comment (acvictor) [reply velox/dwio/common/tests/ReadFileInputStreamTests.cpp]: Done.

File: velox/dwio/common/tests/ReadFileInputStreamTests.cpp (hunk 6 above)
Comment (acvictor) [reply velox/dwio/common/tests/ReadFileInputStreamTests.cpp]: Renamed.

File: velox/dwio/common/tests/ReadFileInputStreamTests.cpp (hunk 5 above)
Comment (acvictor) [reply velox/dwio/common/tests/ReadFileInputStreamTests.cpp]: Done.

File: velox/common/file/File.cpp (hunk 3 above)
Comment (acvictor) [reply velox/common/file/File.cpp]: Removed.

File: velox/common/file/File.cpp (hunk 1 above)
Comment (acvictor) [reply velox/common/file/File.cpp]: Renamed.

**Comment by acvictor** [comment]
Thanks @xiaoxmeng for the review! I have addressed your comments.

File: velox/common/file/File.cpp [hunk 10]
```diff
@@ -142,21 +148,54 @@ uint64_t LocalReadFile::preadv(
...
+  };
+
   for (auto& range : buffers) {
     if (!range.data()) {
       auto skipSize = range.size();
       while (skipSize) {
         auto bytes = std::min<size_t>(droppedBytes.size(), skipSize);
+
+        if (iovecs.size() >= IOV_MAX) {
+          ssize_t bytesRead{0};
```
Comment (xiaoxmeng) [inline velox/common/file/File.cpp:175]: You can just put ssize_t bytesRead{0}; right above the for loop and declare there once?

**Review by xiaoxmeng — APPROVED** [review]
@acvictor LGTM. Thanks!

**Pull request closed without merging**